commit 0aa5f40eea73c67fe186df073a0d4084f1259cbd
Author: bertin <macportsuser@macports.org>
Date:   Sun May 25 22:21:56 2025 +0200

    /opt/local/site-ports/sysutils/MacPorts/files/lamarck/patch_sysctlbyname_emu.diff

diff --git a/src/macports1.0/macports.tcl b/src/macports1.0/macports.tcl
index 9153f889b..1c9fcb0b6 100644
--- a/src/macports1.0/macports.tcl
+++ b/src/macports1.0/macports.tcl
@@ -1365,6 +1365,19 @@ Please edit sources.conf and change '$url' to '[string range $url 0 26]macports/
             if {[catch {file mkdir $portdbpath} result]} {
                 return -code error "portdbpath $portdbpath does not exist and could not be created: $result"
             }
+        } elseif {$os_platform eq "linux"} {
+            try {
+                if {[string match *64-bit* [exec lscpu]]} {
+                    ui_msg "build_arch x86_64"
+                    set macports::build_arch x86_64
+                } else {
+                    ui_msg "build_arch i386"
+                    set macports::build_arch i386
+                }
+            } catch {{*} eCode eMessage} {
+                ui_debug "Error setting macports::build_arch: $eMessage"
+                set macports::build_arch {}
+            }
         } else {
             return -code error "$portdbpath is not a directory. Please create the directory $portdbpath and try again"
         }
diff --git a/src/port1.0/portbuild.tcl b/src/port1.0/portbuild.tcl
index 52dcbbbbd..63e6d3629 100644
--- a/src/port1.0/portbuild.tcl
+++ b/src/port1.0/portbuild.tcl
@@ -150,7 +150,11 @@ proc portbuild::build_getjobs {args} {
     # if set to '0', use the number of cores for the number of jobs
     if {$jobs == 0} {
         macports_try -pass_signal {
-            set jobs [sysctl hw.activecpu]
+            if {[option os.platform] eq "linux"} {
+                set jobs [exec nproc]
+            } else {
+                set jobs [sysctl hw.activecpu]
+            }
         } on error {} {
             set jobs 2
             ui_warn "failed to determine the number of available CPUs (probably not supported on this platform)"
@@ -158,7 +162,12 @@ proc portbuild::build_getjobs {args} {
         }
 
         macports_try -pass_signal {
-            set memsize [sysctl hw.memsize]
+            if {[option os.platform] eq "linux"} {
+                # /proc/meminfo gives the (free!) total memory in kB; we want bytes like `sysctl hw.memsize` gives:
+                set memsize [expr [exec sed -e "s|MemTotal:\[ \]*||g" -e "s| kB.*||g" -e 1q /proc/meminfo] * 1024]
+            } else {
+                set memsize [sysctl hw.memsize]
+            }
             global build.mem_per_job
             set jobs_limit_mem [expr {int($memsize / (${build.mem_per_job} * 1024 * 1024)) + 1}]
             if {$jobs > $jobs_limit_mem} {
